lib/src/session.dart

```diff
@@ -36,12 +87,169 @@ class Session with ChangeNotifier {
   /// code.
   final String? signature;
 
-  final bool enablePushNotifications;
+  HeadlessInAppWebView? _headlessWebView;
+  InAppWebViewController? _webViewController;
+  Completer<void> _completer;
+
+  final bool? enablePushNotifications;
+
+  final MessageHandler? onMessage;
+  final Unreads? unreads;
+
+  void _onWebViewCreated(InAppWebViewController controller) async {
+    if (kDebugMode) {
+      print('📗 session._onWebViewCreated');
+    }
+
+    if (onMessage != null) {
+      controller.addJavaScriptHandler(
+          handlerName: 'JSCOnMessage', callback: _jscOnMessage);
+    }
+
+    if ((unreads != null) && (unreads!.onChange != null)) {
+      controller.addJavaScriptHandler(
+          handlerName: 'JSCOnUnreadsChange', callback: _jscOnUnreadsChange);
+    }
+
+    String htmlData = await rootBundle
+        .loadString('packages/talkjs_flutter/assets/index.html');
+    controller.loadData(
+        data: htmlData, baseUrl: WebUri("https://app.talkjs.com"));
+  }
+
+  void _onLoadStop(InAppWebViewController controller, WebUri? url) async {
+    if (kDebugMode) {
+      print('📗 session._onLoadStop ($url)');
+    }
+
+    if (_webViewController == null) {
+      _webViewController = controller;
+
+      // Wait for TalkJS to be ready
+      final js = 'await Talk.ready;';
+
+      if (kDebugMode) {
+        print('📗 session callAsyncJavaScript: $js');
+      }
+
+      await controller.callAsyncJavaScript(functionBody: js);
+
+      // If the `me` property has already been initialized, then create the user and the session
+      if ((_me != null) && (!_completer.isCompleted)) {
+        _execute('const me = new Talk.User(${me.getJsonString()});');
+        createSession(
+          execute: _execute,
+          session: this,
+          variableName: 'me',
+        );
+
+        if (onMessage != null) {
+          _execute(
+              'session.onMessage((event) => window.flutter_inappwebview.callHandler("JSCOnMessage", JSON.stringify(event)));');
+        }
+
+        if ((unreads != null) && (unreads!.onChange != null)) {
+          _execute(
+              'session.unreads.onChange((event) => window.flutter_inappwebview.callHandler("JSCOnUnreadsChange", JSON.stringify(event)));');
+        }
 
-  Session(
-      {required this.appId,
-      this.signature,
-      this.enablePushNotifications = false});
+        if (enablePushNotifications != null) {
+          await _setOrUnsetPushRegistration(enablePushNotifications!);
+        }
+
+        _completer.complete();
```

Comment:
not really.
I thought about that, but there's a crucial difference:
the `me` setter, being a setter cannot be async (and even if it could, it would be strange to `await` setting a property).
The `await _setOrUnsetPushRegistration` is here because we can complete the `_completer.future` after we're sure that the push registrations have been (un)set.
This was the famous race condition that we had during Mohamed's testing.

This also means that we potentially have a race condition when setting the `me` property and changing the push registrations.
Unfortunately I have no idea on how to solve this.

Reply:
Noted.